documents/entity/model.go

```diff
@@ -0,0 +1,290 @@
+func (e *Entity) JSON() ([]byte, error) {
+	return json.Marshal(e)
+}
+
+// FromJSON unmarshals the json bytes into Entity
+func (e *Entity) FromJSON(jsonData []byte) error {
+	return json.Unmarshal(jsonData, e)
+}
+
+// Type gives the Entity type
+func (e *Entity) Type() reflect.Type {
+	return reflect.TypeOf(e)
+}
+
+// CalculateDataRoot calculates the data root and sets the root to core document.
+func (e *Entity) CalculateDataRoot() ([]byte, error) {
+	t, err := e.getDocumentDataTree()
+	if err != nil {
+		return nil, errors.New("failed to get data tree: %v", err)
+	}
+
+	dr := t.RootHash()
+	e.CoreDocument.SetDataRoot(dr)
+	return dr, nil
+}
+
+// getDocumentDataTree creates precise-proofs data tree for the model
+func (e *Entity) getDocumentDataTree() (tree *proofs.DocumentTree, err error) {
+	entityProto := e.createP2PProtobuf()
+	salts, err := e.getEntitySalts(entityProto)
+	if err != nil {
+		return nil, err
+	}
+	t := documents.NewDefaultTreeWithPrefix(salts, prefix, compactPrefix())
+	err = t.AddLeavesFromDocument(entityProto)
+	if err != nil {
+		return nil, errors.New("getDocumentDataTree error %v", err)
+	}
+	err = t.Generate()
+	if err != nil {
+		return nil, errors.New("getDocumentDataTree error %v", err)
+	}
+	return t, nil
+}
+
+// CreateProofs generates proofs for given fields.
+func (e *Entity) CreateProofs(fields []string) (proofs []*proofspb.Proof, err error) {
+	tree, err := e.getDocumentDataTree()
+	if err != nil {
+		return nil, errors.New("createProofs error %v", err)
+	}
+
+	return e.CoreDocument.CreateProofs(e.DocumentType(), tree, fields)
+}
+
+// DocumentType returns the entity document type.
+func (*Entity) DocumentType() string {
+	return documenttypes.EntityDataTypeUrl
+}
+
+// PrepareNewVersion prepares new version from the old entity.
+func (e *Entity) PrepareNewVersion(old documents.Model, data *cliententitypb.EntityData, collaborators []string) error {
+	err := e.initEntityFromData(data)
+	if err != nil {
+		return err
+	}
+
+	oldCD := old.(*Entity).CoreDocument
+	e.CoreDocument, err = oldCD.PrepareNewVersion(collaborators, true, compactPrefix())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddNFT adds NFT to the Entity.
+func (e *Entity) AddNFT(grantReadAccess bool, registry common.Address, tokenID []byte) error {
+	cd, err := e.CoreDocument.AddNFT(grantReadAccess, registry, tokenID)
+	if err != nil {
+		return err
+	}
+
+	e.CoreDocument = cd
+	return nil
+}
+
+// CalculateSigningRoot calculates the signing root of the document.
+func (e *Entity) CalculateSigningRoot() ([]byte, error) {
+	return e.CoreDocument.CalculateSigningRoot(e.DocumentType())
+}
+
+// CreateNFTProofs creates proofs specific to NFT minting.
+func (e *Entity) CreateNFTProofs(
+	account identity.DID,
+	registry common.Address,
+	tokenID []byte,
+	nftUniqueProof, readAccessProof bool) (proofs []*proofspb.Proof, err error) {
+	return e.CoreDocument.CreateNFTProofs(
+		e.DocumentType(),
```

Comment:
tree?

Reply:
As @charleenfei pointed out, we need to pass the data tree here, right?